pyEpiabm/pyEpiabm/sweep/host_progression_sweep.py

```diff
@@ -119,18 +89,64 @@ def _update_next_infection_status(self, person: Person):
-        outcomes = range(1, self.number_of_states + 1)
-
-        if len(weights) != len(outcomes):
-            raise AssertionError('The number of infection statuses must \
-                                match the number of transition probabilities')
-
-        next_infection_status_number = random.choices(outcomes, weights)[0]
-        next_infection_status = InfectionStatus(next_infection_status_number)
-        person.next_infection_status = next_infection_status
+        if person.infection_status.name in ['Recovered', 'Dead']:
+            person.next_infection_status = None
+        else:
+            row_index = person.infection_status.name
+            weights = self.state_transition_matrix.loc[row_index].to_numpy()
+            outcomes = range(1, self.number_of_states + 1)
+
+            if len(weights) != len(outcomes):
+                raise AssertionError('The number of infection statuses must \
+                                    match the number of transition \
+                                    probabilities')
+
+            next_infection_status_number = random.choices(outcomes, weights)[0]
+            next_infection_status =\
+                InfectionStatus(next_infection_status_number)
+            person.next_infection_status = next_infection_status
+
+    def _update_time_status_change(self, person, time):
+        """Calculates transition time as calculated in CovidSim,
+        and updates the time_of_status_change for the given
+        Person, given as the time until next infection status
+        for a person who has a new infection status. If it is expected that
+        the person will not transition again (for example in Recovered or Dead
+        statuses), then the time of status change is set to infinity.
+
+        :param Person: Person class with infection status attributes
+        :type Person: Person
+        :param time: Current simulation time
+        :type time: float
+        """
+        # Defines the transition time. If the person will not transition again,
+        # the transition time is set to infinity. Else, the transition time is
+        # defined using the TransitionTimeMatrix class, with a method choose
+        # from the InverseCdf class.
+        if (person.infection_status == InfectionStatus.Recovered or
+                person.infection_status == InfectionStatus.Dead):
+            transition_time = np.inf
+        else:
+            row_index = person.infection_status.name
+            column_index = person.next_infection_status.name
+            transition_time_icdf_object =\
+                self.transition_time_matrix.loc[row_index, column_index]
+            # Checks for susceptible to exposed case
+            #  where transition time is zero
+            if transition_time_icdf_object == 0.0:
+                transition_time = 0.0
+            else:
+                transition_time =\
+                    transition_time_icdf_object.icdf_choose_noexp()
```

Comment:
```suggestion
            try:
                transition_time =\
                    transition_time_icdf_object.icdf_choose_noexp()
            except AttributeError as e:
                   if "object has no attribute 'icdf_choose_noexp'" in str(e):
                       transition_time = transition_time_icdf_object
                   else:
                       raise
```
This deserves more thought than I have given it but it may be a more graceful way to handle this.
It also has the advantage that if we want to run some simulation with transitions for which we want a single time, rather than a distribution, this allows us to just put those numbers in the matrix rather than having to contrive some delta function icdf.